File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/SpawnedThreadsIterator.java
package io.littlehorse.sdk.wfsdk.internal;

import io.littlehorse.sdk.common.proto.TypeDefinition.DefinedTypeCase;
import io.littlehorse.sdk.common.proto.VariableType;
import io.littlehorse.sdk.common.proto.WaitForThreadsNode;
import io.littlehorse.sdk.common.proto.WaitForThreadsStrategy;
import io.littlehorse.sdk.wfsdk.SpawnedThreads;

public class SpawnedThreadsIterator implements SpawnedThreads {

    private final WfRunVariableImpl internalStartedThreadVar;

    public SpawnedThreadsIterator(final WfRunVariableImpl internalStartedThreadVar) {
        this.internalStartedThreadVar = internalStartedThreadVar;
        if (internalStartedThreadVar.getTypeDef().getDefinedTypeCase() != DefinedTypeCase.PRIMITIVE_TYPE
                || internalStartedThreadVar.getTypeDef().getPrimitiveType() != VariableType.JSON_ARR) {
            throw new IllegalArgumentException("Only support for json arrays");
        }
    }

    @Override
    public WaitForThreadsNode buildNode(WaitForThreadsStrategy strategy) {
        WaitForThreadsNode.Builder waitNode = WaitForThreadsNode.newBuilder();
        waitNode.setThreadList(BuilderUtil.assignVariable(internalStartedThreadVar));
        waitNode.setStrategy(strategy);
        return waitNode.build();
    }
}

File: sdk-dotnet/LittleHorse.Sdk/Workflow/Spec/SpawnedThreadsIterator.cs
using LittleHorse.Sdk.Common.Proto;

namespace LittleHorse.Sdk.Workflow.Spec;

/// <summary>
/// Represents a collection of spawned threads that are fixed at the time of creation.
/// </summary>
public class SpawnedThreadsIterator: SpawnedThreads
{
    private readonly WfRunVariable _internalStartedThreadVar;

    /// <summary>
    /// Initializes a new instance of the <see cref="SpawnedThreadsIterator"/> class with the specified spawned threads.
    /// </summary>
    /// <param name="internalStartedThreadVar">This is a wfRunVariable of VariableType.JsonArr type</param>
    /// <exception cref="ArgumentException">This exception is raised when the _internalStartedThreadVar parameter
    /// does not have a type of VariableType.JsonArr.</exception>
    public SpawnedThreadsIterator(WfRunVariable internalStartedThreadVar) 
    {
       _internalStartedThreadVar = internalStartedThreadVar;
        if (_internalStartedThreadVar.Type != VariableType.JsonArr) 
        {
            throw new ArgumentException("Only support for json arrays.");
        }
    }
    
    /// <summary>
    /// Builds a <see cref="WaitForThreadsNode"/> that represents the spawned threads.
    /// </summary>
    /// <returns>WaitForThreadsNode</returns>
    public WaitForThreadsNode BuildNode(WaitForThreadsStrategy strategy) 
    {
        var variableAssignment = new VariableAssignment();
        if (_internalStartedThreadVar.JsonPath != null) 
        {
            variableAssignment.JsonPath = _internalStartedThreadVar.JsonPath;
        }
        variableAssignment.VariableName = _internalStartedThreadVar.Name;
        
        var waitNode = new WaitForThreadsNode
        {
            ThreadList = variableAssignment,
            Strategy = strategy
        };
        
        return waitNode;
    }
}